Add ERC: Native ETH embedding in NFTs

ten-io-meta opened the pull request:
This PR introduces IERC721Value, a proposed ERC extension that enables embedding native ETH directly within ERC-721 tokens.

It links artistic and cultural assets with real ETH collateral, allowing NFTs to act as verifiable stores of embedded on-chain value.

Discussion thread: https://ethereum-magicians.org/t/draft-erc-ierc721value-embedding-native-eth-inside-erc-721-tokens/25979  
Reference implementation: https://github.com/ten-io-meta/TENIO-Fragment

ten-io-meta committed:
Add draft ERC: IERC721Value — Embedding native ETH inside ERC-721 tokens

ten-io-meta commented:
Author’s note

This proposal is not theoretical — it is based on a verified mainnet deployment.

IERC721Value originates from the TEN.IO project, which introduced the first NFT contract on Ethereum Mainnet with embedded native ETH collateral and reversible burn mechanics.

The implementation has been publicly verified since June 28, 2025, under contract address
[0x7c88fabe05fd373d56a728f9896f87ede10db789](https://etherscan.io/address/0x7c88fabe05fd373d56a728f9896f87ede10db789)
.

This ERC formalizes that mechanism as a general standard for verifiable ETH-backed cultural assets — connecting artistic meaning with on-chain economic energy.

Reference repository: [TEN.IO-Fragment](https://github.com/ten-io-meta/TENIO-Fragment)

ten-io-meta commented:
Technical rationale

The IERC721Value interface introduces a direct native-ETH embedding layer within ERC-721 tokens.

Each mint transaction carries real ETH as collateral (msg.value), which becomes permanently bound to the tokenId within the contract storage.

The contract ensures:

1:1 verifiable balance: valueOf(tokenId) always matches the ETH held on-chain.

Reversible collateralization: When a holder executes burn(tokenId), the token is destroyed and the embedded ETH is released back to the same address.

Non-inflationary design: No additional supply or synthetic wrapping — ETH remains native, not represented as an ERC-20.

This architecture enables NFTs to serve as programmable ETH reservoirs — verifiable stores of both artistic and economic energy.
It maintains ERC-721 compatibility and can integrate seamlessly into DeFi or cultural registries without external or custodial layers.

ten-io-meta committed:
Update ERCS/draft_ierc721value.md

Co-authored-by: Andrew B Coathup <28278242+abcoathup@users.noreply.github.com>

ten-io-meta committed:
Update ERCS/draft_ierc721value.md

Co-authored-by: Andrew B Coathup <28278242+abcoathup@users.noreply.github.com>

ten-io-meta committed:
Update ERCS/draft_ierc721value.md

Co-authored-by: Andrew B Coathup <28278242+abcoathup@users.noreply.github.com>

ten-io-meta committed:
Rename ERCS/draft_ierc721value.md to ERCS/ERCS/eip-8060.md

ten-io-meta committed:
Merge branch 'ethereum:master' into master

ten-io-meta committed:
Merge branch 'master' into master

ten-io-meta committed:
Update eip-8060.md

ten-io-meta committed:
Rename eip-8060.md to erc-8060.md

ten-io-meta committed:
Update erc-8060.md

ten-io-meta committed:
Update erc-8060.md

ten-io-meta committed:
Update erc-8060.md

ten-io-meta committed:
Update erc-8060.md

ten-io-meta committed:
Update erc-8060.md

ten-io-meta committed:
EIPS/eip-8060.md

ten-io-meta committed:
Update eip-8060.md

ten-io-meta committed:
Rename ERCS/ERCS/EIPS/eip-8060.md to ERCS/erc-8060.md

ten-io-meta committed:
Update erc-8060.md

ten-io-meta committed:
Update erc-8060.md

ten-io-meta committed:
Update erc-8060.md

ten-io-meta committed:
Update erc-8060.md

ten-io-meta committed:
Update erc-8060.md

ten-io-meta committed:
Update erc-8060.md

ten-io-meta committed:
Update erc-8060.md

ten-io-meta committed:
Update erc-8060.md

ten-io-meta committed:
Merge branch 'master' into master

ten-io-meta commented:
Update: Value clarification and live implementation

Following feedback, I added a clarification on Ethereum Magicians:
https://ethereum-magicians.org/t/erc-8060-ierc721value-embedding-native-eth-inside-erc-721-tokens/20971/5

Key points:

• Fragment 0 ("El Umbral") mints at 0.012 ETH and holds 0.010 ETH as native collateral.
• Future fragments will follow the structure 0.12 ETH → 0.10 ETH as collateral.
• Full implementation deployed and verified (Sepolia + Mainnet).

Ready for further review. ✅

ten-io-meta committed:
Merge branch 'master' into master

SamWilsn commented on ERCS/erc-8060.md:
```diff
@@ -0,0 +1,33 @@
+This ERC extends [ERC-721](./eip-721.md) and adds:
+- `valueOf(uint256 tokenId) → uint256`: returns the ETH embedded for a token.
+- `mintWithValue(address to, uint256 tokenId)` (**payable**): mints a token embedding `msg.value`.
+- `burn(uint256 tokenId)`: burns a token and refunds the embedded ETH to the owner.
```
This is fine for a draft, but I think it could use a bit more fleshing out.

For example, who is allowed to call `burn`? Can I burn someone else's token?

You'll want to imagine someone implementing this. What questions would they ask? What potential incompatibilities could crop up?

SamWilsn commented on ERCS/erc-8060.md:
```diff
@@ -0,0 +1,33 @@
+- `burn(uint256 tokenId)`: burns a token and refunds the embedded ETH to the owner.
+
+## Rationale
+The design preserves ERC-721 compatibility while enabling per-token ETH balances as verifiable, on-chain collateral.
```
The Rationale section should be used to explain choices made within the proposal itself, while the Motivation justifies the proposal as a whole. My favourite analogy:

> Motivation: We need to build a shed because...
> Rationale: We chose steel doors for the shed because...

ten-io-meta committed:
Expand Motivation section with use cases and interested actors

ten-io-meta committed:
Expand Rationale section to clarify design decisions

ten-io-meta committed:
Fix copyright waiver to match EIP template

ten-io-meta committed:
Clarify burn-based ETH redemption semantics

ten-io-meta committed:
Fix copyright section to match EIP template

ten-io-meta committed:
Merge branch 'master' into master

ten-io-meta committed:
Merge branch 'master' into master

ten-io-meta committed:
Refine ERC-8060 specification and expand motivation

- Expanded Motivation with concrete use cases and actors
- Clarified burn authorization (owner-only redemption)
- Introduced valueOf(tokenId) as a minimal interface for redeemable value
- Made specification implementation-agnostic
- Improved Rationale separation and technical clarity
- Added detailed security considerations

ten-io-meta committed:
Refine ERC-8060 specification and address review feedback

Addressed remaining formatting and security clarifications:

- fixed CC0 license formatting to match EIP template
- clarified burn authorization semantics
- expanded Security Considerations (reentrancy, accounting, access control)

Happy to iterate further.

ten-io-meta committed:
Refine ERC-8060 specification and address review feedback

- Clarify IERC721Value interface and redemption semantics.
- Expand Motivation with use cases and interested actors.
- Add Value Assignment, Accounting Model and ETH Balance Considerations.
- Clarify owner-only redemption behavior.
- Expand Security Considerations and align copyright waiver with EIP template.
- Preserve full ERC-721 compatibility while remaining implementation-agnostic.

ten-io-meta committed:
Fix EIP validator markdown-link-first requirement

ten-io-meta committed:
Fix markdown ERC reference validation

ten-io-meta committed:
Merge branch 'master' into ten-io-meta-patch-1

ten-io-meta committed:
Merge pull request #1 from ten-io-meta/ten-io-meta-patch-1

Refine ERC-8060 spec: value interface, motivation and security clarifications

SamWilsn commented on ERCS/erc-8060.md:
```diff
@@ -0,0 +1,175 @@
+type: Standards Track
+category: ERC
+created: 2025-10-26
+requires: 721
```
```suggestion
requires: 165, 721
```

SamWilsn commented on ERCS/erc-8060.md:
```diff
@@ -0,0 +1,175 @@
+- MUST ensure that the redeemed value cannot be claimed again.
+- MUST emit the standard ERC-721 `Transfer` event to the zero address as required for burns.
+
+Implementations MAY define additional minting functions, payable minting flows, or value-assignment mechanisms. These are outside the scope of the required interface.
```
It's generally assumed that, unless explicitly forbidden, contracts can add whatever other functions they want.

ten-io-meta replied:
Thanks, removed the redundant sentence.

SamWilsn commented on ERCS/erc-8060.md:
```diff
@@ -0,0 +1,175 @@
+
+Implementations SHOULD define how surplus ETH is handled when the contract balance exceeds total redeemable obligations.
+
+Failure to properly manage authorization, accounting, or redemption logic may result in insolvency, unintended loss of funds, or unauthorized value extraction.
```
Don't create requirements (defined with UPPERCASE keywords) outside of the Specification section. It's too easy for implementers to miss them here.

That said, you should absolutely discuss security implications here.

ten-io-meta replied:
Thanks, updated.

I revised the Security Considerations section to discuss the relevant risks without introducing RFC2119 requirements outside the Specification section.

ten-io-meta committed:
Update ERCS/erc-8060.md

Applied, thanks.

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>

ten-io-meta committed:
docs: address review feedback and clarify burn semantics

Address reviewer feedback by clarifying burn semantics, moving normative requirements into the Specification section, and refining the rationale around token lifecycle and redemption guarantees. No functional changes to the ERC.
